modules/fedimint-wallet-client/src/lib.rs

```diff
@@ -127,61 +172,185 @@ impl WalletClientExt for Client {
+    }
+
+    async fn subscribe_withdraw_updates(
+        &self,
+        operation_id: OperationId,
+    ) -> anyhow::Result<UpdateStreamOrOutcome<WithdrawState>> {
+        let (wallet_client, _) =
+            self.get_first_module::<WalletClientModule>(&WalletCommonGen::KIND);
+
+        let operation = self
+            .get_operation(operation_id)
+            .await
+            .ok_or(anyhow!("Operation not found"))?;
+
+        if operation.operation_type() != WalletCommonGen::KIND.as_str() {
+            bail!("Operation is not a wallet operation");
+        }
+
+        let operation_meta = operation.meta::<WalletOperationMeta>();
+
+        let WalletOperationMeta::Withdraw { change, .. } = operation_meta else {
+            bail!("Operation is not a withdraw operation");
+        };
+
+        let mut operation_stream = wallet_client.notifier.subscribe(operation_id).await;
+
+        Ok(
+            operation.outcome_or_updates(self.db(), operation_id, move || {
+                stream! {
+                    match next_withdraw_state(&mut operation_stream).await {
+                        Some(WithdrawStates::Created(_)) => {
+                            yield WithdrawState::Created;
+                        },
+                        Some(s) => {
+                            panic!("Unexpected state {s:?}")
+                        },
+                        None => return,
+                    }
+
+                    // TODO: get rid of awaiting change here, there has to be a better way to make tests deterministic
+                    if let Some(change_out_point) = change {
+                        // Swallowing potential errors since the transaction failing  is handled by
+                        // output outcome fetching already
+                        let _ = self
+                            .await_primary_module_output(operation_id, change_out_point)
+                            .await;
+                    }
+
+                    match next_withdraw_state(&mut operation_stream).await {
+                        Some(WithdrawStates::Aborted(inner)) => {
+                            yield WithdrawState::Failed(inner.error);
+                        },
+                        Some(WithdrawStates::Success(inner)) => {
+                            yield WithdrawState::Succeeded(inner.txid);
+                        },
+                        Some(s) => {
+                            panic!("Unexpected state {s:?}")
+                        },
+                        None => {},
+                    }
+                }
+            }),
+        )
     }
 }
 
 async fn next_deposit_state<S>(stream: &mut S) -> Option<DepositStates>
 where
     S: Stream<Item = WalletClientStates> + Unpin,
 {
-    match stream.next().await? {
-        WalletClientStates::Deposit(ds) => Some(ds.state),
+    loop {
+        if let WalletClientStates::Deposit(ds) = stream.next().await? {
+            return Some(ds.state);
+        }
```

Comment:
I like to use [yield_now](https://docs.rs/tokio/latest/tokio/task/fn.yield_now.html) inside loops on async functions. It avoids starving other threads:
```suggestion
        }
        tokio::task::yield_now().await;
```

It saved me a couple of times.

Reply:
TIL, added it, but want to get rid of these functions anyway, not entirely sure how yet.